### PR TITLE
Added PresetCookieHeader method that sets cookie token and header

### PR DESCRIPTION
**Pull request by kirthik**
@yishaigalatzer, @harshgMSFT, @DamianEdwards  - Could you please take a look?

### CODE REVIEW

**Comment by yishaigalatzer** [comment]
From quick reading over the phone. I got a couple of comments:

1. The new method should no be on the html helper.
2. The method names should be consistent SetCookieAndHeader on an AntiForgery helper. The name getcookie doesn't fit in a method returning void

**Comment by kirthik** [comment]
Ok. I will change method names all across to SetCookieAndHeader. Where should the new method be if not on HtmlHelper?

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 1]
```diff
@@ -204,6 +192,50 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
                 deserializedFormToken);
         }
 
+        // [ ENTRY POINT ]
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: public API need XML docs, and not just // comments

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 2]
```diff
@@ -204,6 +192,50 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
...
+
+        private AntiForgeryToken ValidateAndGenerateNewToken(AntiForgeryToken oldCookieToken)
+        {
+            AntiForgeryToken newCookieToken = null;
+
+            if (!_validator.IsCookieTokenValid(oldCookieToken))
+            {
+                // Need to make sure we're always operating with a good cookie token.
+                newCookieToken = _generator.GenerateCookieToken();
+            }
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: new line after closing } (and below)

File: src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs [hunk 3]
```diff
@@ -82,6 +82,12 @@ public interface IHtmlHelper
         HtmlString AntiForgeryToken();
 
         /// <summary>
+        /// Returns anti-forgery cookie token and header.
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs]: This should say sets the cookie and X-Frame-Options header on the response.

Add remarks for the use case of flush async

File: src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs [hunk 4]
```diff
@@ -82,6 +82,12 @@ public interface IHtmlHelper
...
+        /// </summary>
+        /// <returns>A<see cref="HtmlString"/> that returns a <see cref="HtmlString.Empty"/>.</returns>
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/IHtmlHelper.cs]: `Returns a <see cref="HtmlString.Empty" />.`

File: test/Microsoft.AspNet.Mvc.Core.Test/AntiXsrf/AntiForgeryWorkerTest.cs [hunk 5]
```diff
@@ -379,6 +379,29 @@ public void Validate_FromValidStrings_TokensValidatedSuccessfully()
             context.TokenProvider.Verify();
         }
```
Comment (yishaigalatzer) [inline test/Microsoft.AspNet.Mvc.Core.Test/AntiXsrf/AntiForgeryWorkerTest.cs]: Add a test that makes sure interaction between `Preset`, `FormTagHelper` and `AntiForgeyToken` works as expected (cookies are not overriden, POST works fine)

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs [hunk 6]
```diff
@@ -97,5 +97,18 @@ public void Validate([NotNull] HttpContext context, AntiForgeryTokenSet antiForg
...
+
+        /// <summary>
+        /// Generates an anti-forgery cookie token for this request. 
+        /// </summary>
+        /// <param name="context">The HTTP context associated with the current call.</param>
+        /// <remarks>
+        /// This method has a side effect:
+        /// A response cookie is set if there is no valid cookie associated with the request.
+        /// </remarks>
+        public void GetCookieTokenAndHeader([NotNull] HttpContext context)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs]: Should be something like ```SetCookieAndTokenHeader``` there is nothing to get.
Also remove the remark.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 7]
```diff
@@ -143,15 +135,11 @@ public AntiForgeryTokenSet GetTokens([NotNull] HttpContext httpContext, string s
 
         private AntiForgeryTokenSetInternal GetTokens(HttpContext httpContext, AntiForgeryToken oldCookieToken)
         {
-            AntiForgeryToken newCookieToken = null;
-            if (!_validator.IsCookieTokenValid(oldCookieToken))
+            AntiForgeryToken newCookieToken = ValidateAndGenerateNewToken(oldCookieToken);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: var

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 8]
```diff
@@ -204,6 +192,50 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
...
+                newCookieToken = _generator.GenerateCookieToken();
+            }
+            if (newCookieToken != null)
+            {
+                Debug.Assert(_validator.IsCookieTokenValid(newCookieToken));
+            }
+            return newCookieToken;
+        }
+
+        private void SaveCookieTokenAndHeader([NotNull] HttpContext httpContext, [NotNull] AntiForgeryToken newCookieToken)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: line break.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 9]
```diff
@@ -204,6 +192,50 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
...
+        {
+            // If a new cookie was generated, persist it.
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: I do nt see an if here

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 10]
```diff
@@ -204,6 +192,50 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
                 deserializedFormToken);
         }
 
+        // [ ENTRY POINT ]
+        // Generates an anti-XSRF cookie token for the current user.
+        public void GetCookieTokenAndHeader([NotNull] HttpContext httpContext)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: update the name ... ```SetCookieTokenAndHeader```

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 11]
```diff
@@ -204,6 +192,50 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
...
+        {
+            CheckSSLConfig(httpContext);
+
+            var oldCookieToken = GetCookieTokenNoThrow(httpContext);
+            AntiForgeryToken newCookieToken = ValidateAndGenerateNewToken(oldCookieToken);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: var

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 12]
```diff
@@ -204,6 +192,50 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
...
+            if (newCookieToken != null)
+            {
+                SaveCookieTokenAndHeader(httpContext, newCookieToken);
+            }
+        }
+
+        private AntiForgeryToken ValidateAndGenerateNewToken(AntiForgeryToken oldCookieToken)
+        {
+            AntiForgeryToken newCookieToken = null;
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: if the old cookie is still valid, wouldn't we always return null from this method?

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 13]
```diff
@@ -204,6 +192,50 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
...
+
+            if (!_validator.IsCookieTokenValid(oldCookieToken))
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: Couldn't this be rewritten as 
```C#
if (!_validator.IsCookieTokenValid(oldCookieToken))
{
   return _generator.GenerateCookieToken();
}

return null;
```

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs (hunk 12 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: yes that is correct. Is that an issue?

Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: ah we are doing a null check, please add a comment to the method saying that this method returns null if the old cookie is still valid.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/IHtmlGenerator.cs [hunk 14]
```diff
@@ -31,6 +31,8 @@ public interface IHtmlGenerator
 
         TagBuilder GenerateAntiForgery([NotNull] ViewContext viewContext);
 
+        void GeneratePresetCookieAndHeaders([NotNull] ViewContext viewContext);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/IHtmlGenerator.cs]: Doc comments.

File: test/Microsoft.AspNet.Mvc.Core.Test/AntiXsrf/AntiForgeryWorkerTest.cs [hunk 15]
```diff
@@ -379,6 +379,29 @@ public void Validate_FromValidStrings_TokensValidatedSuccessfully()
...
+
+            // Genreate a new cookie.
+            var context = GetAntiForgeryWorkerContext(options, useOldCookie: false, isOldCookieValid: false);
+            var worker = GetAntiForgeryWorker(context);
+
+            // Act
+            worker.GetCookieTokenAndHeader(context.HttpContext.Object);
+
+            // Assert
+            string xFrameOptions = context.HttpContext.Object.Response.Headers["X-Frame-Options"];
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Core.Test/AntiXsrf/AntiForgeryWorkerTest.cs]: `var`

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultHtmlGenerator.cs [hunk 16]
```diff
@@ -87,6 +87,12 @@ public virtual TagBuilder GenerateAntiForgery([NotNull] ViewContext viewContext)
         }
 
         /// <inheritdoc />
+        public virtual void GeneratePresetCookieAndHeaders([NotNull] ViewContext viewContext)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultHtmlGenerator.cs]: mm probably does nt need to be an HTML helpr since this is not generating any html code, probably better off as a separate static AF helper or something.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs [hunk 17]
```diff
@@ -234,5 +234,27 @@ public class AntiForgeryTests
...
+
+            // Assert
+            var header = Assert.Single(response.Headers.GetValues("X-Frame-Options"));
+            Assert.Equal("SAMEORIGIN", header);
+
+            var setCookieHeader = response.Headers.GetValues("Set-Cookie").ToArray();
+
+            // Even though there are two forms there should only be one response cookie,
+            // as for the second form, the cookie from the first token should be reused.
+            Assert.Equal(1, setCookieHeader.Length);
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs]:
```c#
var cookie = Assert.Single(setCookieHeader);
Assert.True(cookie.StartsWith("..');
```

File: test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs [hunk 18]
```diff
@@ -234,5 +234,27 @@ public class AntiForgeryTests
...
+            Assert.True(setCookieHeader[0].StartsWith("__RequestVerificationToken"));
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs]: do you also want to verify there are two form tokens?

File: test/WebSites/AntiForgeryWebSite/Views/Shared/_FlushAsyncLayout.cshtml [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+<head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>@ViewBag.Title –AntiForgery Functional Tests</title>
+</head>
+@Html.PresetCookieAndHeaders()
+@await FlushAsync()
+
+<body>
+    <div class="navbar navbar-inverse navbar-fixed-top">
```
Comment (harshgMSFT) [inline test/WebSites/AntiForgeryWebSite/Views/Shared/_FlushAsyncLayout.cshtml]: remove the unnecessary clutter if not needed.

**Comment by harshgMSFT** [comment]
looks in the right direction :watch:  on name changes and the HtmlHelper removal change.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs (hunk 13 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: Yep. Fixed it. Thanks for pointing it out.

File: src/Microsoft.AspNet.Mvc.Core/Rendering/Html/IHtmlGenerator.cs (hunk 14 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/Rendering/Html/IHtmlGenerator.cs]: Waiting for decision on whether this method needs to be here. So will hold off on adding comments until the decision is made.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs (hunk 18 above)
Comment (kirthik) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs]: There are no form tokens here. Comment was a copy-paste mistake.

File: test/Microsoft.AspNet.Mvc.Core.Test/AntiXsrf/AntiForgeryWorkerTest.cs (hunk 5 above)
Comment (kirthik) [inline test/Microsoft.AspNet.Mvc.Core.Test/AntiXsrf/AntiForgeryWorkerTest.cs]: Added a new test for this case

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs [hunk 20]
```diff
@@ -97,5 +97,14 @@ public void Validate([NotNull] HttpContext context, AntiForgeryTokenSet antiForg
         {
             Validate(context, antiForgeryTokenSet.CookieToken, antiForgeryTokenSet.FormToken);
         }
+
+        /// <summary>
+        /// Generates an anti-forgery cookie token for this request.
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs]: Generates and sets an .... if one is not available yet (or something along these lines)

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs [hunk 21]
```diff
@@ -97,5 +97,14 @@ public void Validate([NotNull] HttpContext context, AntiForgeryTokenSet antiForg
         {
             Validate(context, antiForgeryTokenSet.CookieToken, antiForgeryTokenSet.FormToken);
         }
+
+        /// <summary>
+        /// Generates an anti-forgery cookie token for this request. 
+        /// </summary>
+        /// <param name="context">The HTTP context associated with the current call.</param>
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs]: see cref etc.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 22]
```diff
@@ -204,6 +189,54 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
...
+            }
+
+            return null;
+        }
+
+        private void SaveCookieTokenAndHeader(
+            [NotNull] HttpContext httpContext,
+            [NotNull] AntiForgeryToken newCookieToken)
+        {
+            if (newCookieToken != null)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: isnt  this check bogus? we have NotNull already. I understand the need for it but I suggest this check should be where you are calling this method.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 23]
```diff
@@ -204,6 +189,54 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
...
+        /// </summary>
+        /// <param name="context">The HTTP context associated with the current call.</param>
+        public void SetCookieTokenAndHeader([NotNull] HttpContext httpContext)
+        {
+            CheckSSLConfig(httpContext);
+
+            var oldCookieToken = GetCookieTokenNoThrow(httpContext);
+            var newCookieToken = ValidateAndGenerateNewToken(oldCookieToken);
+            
+            SaveCookieTokenAndHeader(httpContext, newCookieToken);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: Do the null check here instead of the method.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs (hunk 22 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: I will remove [NotNull] on newCookieToken. newCookieToken can be null in cases where oldCookieToken is valid. In this case, I don't want to save the cookie but just write X-Frame-Options header to response

Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: or remove the NotNull attribute from  here.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs [hunk 24]
```diff
@@ -204,6 +189,54 @@ public void Validate([NotNull] HttpContext httpContext, string cookieToken, stri
                 deserializedFormToken);
         }
 
+
+        /// <summary>
+        /// Generates an anti-forgery cookie token and header for this request.
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: I suggest drop the Generates as it only generates if it cant find one or the existing token in not valid.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs (hunk 23 above)
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgeryWorker.cs]: This method saves cookie and also writes X-frame-options to response header. So I have to call this method even if newcookie is null. I will remove [NotNull] on newCookieToken parameter.

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 25]
```diff
@@ -39,12 +39,7 @@ public HttpContext Context
         {
             get
             {
-                if (ViewContext == null)
-                {
-                    return null;
-                }
-
-                return ViewContext.HttpContext;
+                return ViewContext?.HttpContext;
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: this is fine but in general avoid unrelated changes.. :smile:

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 26]
```diff
@@ -561,7 +556,9 @@ public Task<HtmlString> RenderSectionAsync([NotNull] string name)
         /// <returns>A<see cref="Task{HtmlString}"/> that represents the asynchronous flush operation and on
         /// completion returns a <see cref="HtmlString.Empty"/>.</returns>
         /// <remarks>The value returned is a token value that allows FlushAsync to work directly in an HTML
-        /// section. However the value does not represent the rendered content.</remarks>
+        /// section. However the value does not represent the rendered content.
+        /// Call <see cref="SetAntiForgeryCookieAndHeader"/> to send anti-forgery cookie token  and and X-Frame-Options
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]:
```This method also writes out headers, any modifications to headers must be done before FulshAsync is called. 
For example Call ... <rest of the comment that you have>```

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 27]
```diff
@@ -617,6 +614,20 @@ public void EndContext()
             PageExecutionContext?.EndContext();
         }
 
+        /// <summary>
+        /// Sets anti-forgery cookie and X-Frame-Options header on the response.
+        /// </summary>
+        /// <returns>A <see cref="HtmlString"/> that returns a <see cref="HtmlString.Empty"/>.</returns>
+        /// <remarks> Call this method to send anti-forgery cookie token and X-Frame-Options header to client
+        /// before <see cref="FlushAsync"/> flushes the headers. </remarks>
+        public virtual HtmlString SetAntiForgeryCookieAndHeader()
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Does this have to return an HtmlString?

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 28]
```diff
@@ -617,6 +614,20 @@ public void EndContext()
             PageExecutionContext?.EndContext();
         }
 
+        /// <summary>
+        /// Sets anti-forgery cookie and X-Frame-Options header on the response.
+        /// </summary>
+        /// <returns>A <see cref="HtmlString"/> that returns a <see cref="HtmlString.Empty"/>.</returns>
+        /// <remarks> Call this method to send anti-forgery cookie token and X-Frame-Options header to client
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: nit extra space before call.

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 29]
```diff
@@ -617,6 +614,20 @@ public void EndContext()
...
+        /// before <see cref="FlushAsync"/> flushes the headers. </remarks>
+        public virtual HtmlString SetAntiForgeryCookieAndHeader()
+        {
+            var antiForgery = Context?.RequestServices.GetRequiredService<AntiForgery>();
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: will the context ever be null? if not remove ?. with .  in both the occurrences in this code.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs [hunk 30]
```diff
@@ -234,5 +234,59 @@ public class AntiForgeryTests
             Assert.Equal("The required anti-forgery form field \"__RequestVerificationToken\" is not present.",
                          exception.ExceptionMessage);
         }
+
+        [Fact]
+        public async Task SetCookieAndHeaderBeforeFlushAsync_GeneratesCookieTokenAndHeader()
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs]: should there be one where we try to set it after ?

File: test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs [hunk 31]
```diff
@@ -234,5 +234,59 @@ public class AntiForgeryTests
...
+
+            var formToken = AntiForgeryTestHelper.RetrieveAntiForgeryToken(resposneBody, "Account/FlushAsyncLogin");
+            var cookieToken = AntiForgeryTestHelper.RetrieveAntiForgeryCookie(getResponse);
+
+            var request = new HttpRequestMessage(HttpMethod.Post, "http://localhost/Account/FlushAsyncLogin");
+            request.Headers.Add("Cookie", "__RequestVerificationToken=" + cookieToken);
+            var nameValueCollection = new List<KeyValuePair<string, string>>
+            {
+                new KeyValuePair<string,string>("__RequestVerificationToken", formToken),
+                new KeyValuePair<string,string>("UserName", "abra"),
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs]: use better names please .. :smile:

File: test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs [hunk 32]
```diff
@@ -234,5 +234,59 @@ public class AntiForgeryTests
...
+            Assert.Equal("SAMEORIGIN", header);
+
+            var setCookieHeader = response.Headers.GetValues("Set-Cookie").ToArray();
+
+            var cookie = Assert.Single(setCookieHeader);
+            Assert.True(cookie.StartsWith("__RequestVerificationToken"));
+        }
+
+        [Fact]
+        public async Task SetCookieAndHeaderBeforeFlushAsync_PostToForm()
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTests.cs]: should there be one for the failure case as well ?

**Comment by harshgMSFT** [comment]
looks almost done :watch: for NotNull and test cases.

**Comment by harshgMSFT** [comment]
:shipit: after fixing the comments.
